### PR TITLE
feat(TDI-50082): Add tcompv0 tck migration jira

### PR DESCRIPTION
**Pull request by dmytro-sylaiev**
**What is the current behavior?** (You can also link to an open issue here)

**What is the new behavior?**

**Please check if the PR fulfills these requirements**

- [ ] The commit message follows Talend standard
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features) ?
- [ ] The code coverage on new code >75%
- [ ] The new code does not introduce new technical issues (sonar / eslint)

**What kind of change does this PR introduce?**

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build / CI related changes
- [ ] Other... Please describe:

**Does this PR introduce a breaking change?**

- [ ] Yes
- [ ] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

### CODE REVIEW

File: main/plugins/org.talend.sdk.component.studio-integration/src/main/java/org/talend/sdk/component/studio/metadata/migration/tcompv0/JiraTcompv0ToTCKMigrationTask.java [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+import org.talend.designer.core.model.utils.emf.talendfile.ElementParameterType;
+import org.talend.designer.core.model.utils.emf.talendfile.NodeType;
+import org.talend.designer.core.model.utils.emf.talendfile.ProcessType;
+import org.talend.sdk.component.studio.util.TcompV0ConfigToFlatMapDeserializer;
+import org.talend.designer.core.model.utils.emf.talendfile.MetadataType;
+import org.talend.designer.core.model.utils.emf.talendfile.ConnectionType;
+
+public class JiraTcompv0ToTCKMigrationTask extends AbstractJobMigrationTask {
+
+    private Item item;
```
Comment (mhryb) [inline main/plugins/org.talend.sdk.component.studio-integration/src/main/java/org/talend/sdk/component/studio/metadata/migration/tcompv0/JiraTcompv0ToTCKMigrationTask.java:30]: Not sure why do we need it, if we don't use it?

File: main/plugins/org.talend.sdk.component.studio-integration/src/main/java/org/talend/sdk/component/studio/metadata/migration/tcompv0/JiraTcompv0ToTCKMigrationTask.java [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+        boolean modified = false;
+        ComponentCategory category = ComponentCategory.getComponentCategoryFromItem(item);
+
+        for (int i = 0; i < componentsName.length; i++) {
+            IComponentFilter filter = new NameComponentFilter(componentsName[i]);
+            String oldComponentName = componentsName[i];
+            final String newComponentName = oldComponentName.substring(1); //tck connector name doesn't start with t
+            try {
+                modified |= ModifyComponentsAction.searchAndModify(item, processType, filter,
+                        Arrays.<IComponentConversion>asList(new IComponentConversion() {
```
Comment (mhryb) [inline main/plugins/org.talend.sdk.component.studio-integration/src/main/java/org/talend/sdk/component/studio/metadata/migration/tcompv0/JiraTcompv0ToTCKMigrationTask.java:51]: I would move it to separate variable for better readability (just my personal opinion)

File: main/plugins/org.talend.sdk.component.studio-integration/src/main/java/org/talend/sdk/component/studio/metadata/migration/tcompv0/JiraTcompv0ToTCKMigrationTask.java [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+                                node.setComponentVersion("1");
+                                ElementParameterType elemParamType =
+                                        ComponentUtilities.getNodeProperty(node, "PROPERTIES");
+
+                                final String jsonProperties = elemParamType.getValue();
+
+                                Map<String, String> tcompv0PropertiesToMap =
+                                        TcompV0ConfigToFlatMapDeserializer.configToMap(jsonProperties);
+
+                                //amlyYSNKSVJBI091dHB1dA
```
Comment (mhryb) [inline main/plugins/org.talend.sdk.component.studio-integration/src/main/java/org/talend/sdk/component/studio/metadata/migration/tcompv0/JiraTcompv0ToTCKMigrationTask.java:64]: Not sure what it means

File: main/plugins/org.talend.sdk.component.studio-integration/src/main/java/org/talend/sdk/component/studio/util/TcompV0ConfigToFlatMapDeserializer.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        JsonReader reader = Json.createReader(new StringReader(jsonConfig));
+        JsonObject configAsJsonObject = reader.readObject();
+
+        Map<String,String> result = new LinkedHashMap<>();
+
+        configAsJsonObject.keySet().stream().filter(key ->
+                configAsJsonObject.get(key) instanceof JsonObject).forEach(key -> {
+            JsonObject value = (JsonObject) configAsJsonObject.get(key);
+            findStoredValueInJson(key, result, value);
+        });
```
Comment (mhryb) [inline main/plugins/org.talend.sdk.component.studio-integration/src/main/java/org/talend/sdk/component/studio/util/TcompV0ConfigToFlatMapDeserializer.java:24]: Just suggestion for more readability
```suggestion
        configAsJsonObject.keySet().stream()
            .filter(key -> configAsJsonObject.get(key) instanceof JsonObject)
            .forEach(key -> {
                JsonObject value = (JsonObject) configAsJsonObject.get(key);
                findStoredValueInJson(key, result, value);
        });
```

File: main/plugins/org.talend.sdk.component.studio-integration/src/main/java/org/talend/sdk/component/studio/metadata/migration/tcompv0/JiraTcompv0ToTCKMigrationTask.java (hunk 2 above)
Comment (dmytro-sylaiev) [reply main/plugins/org.talend.sdk.component.studio-integration/src/main/java/org/talend/sdk/component/studio/metadata/migration/tcompv0/JiraTcompv0ToTCKMigrationTask.java:38]: I don't agree that would improve readability, also I respected the code style with other migration tasks

File: main/plugins/org.talend.sdk.component.studio-integration/src/main/java/org/talend/sdk/component/studio/metadata/migration/tcompv0/JiraTcompv0ToTCKMigrationTask.java (hunk 1 above)
Comment (dmytro-sylaiev) [reply main/plugins/org.talend.sdk.component.studio-integration/src/main/java/org/talend/sdk/component/studio/metadata/migration/tcompv0/JiraTcompv0ToTCKMigrationTask.java]: removed

File: main/plugins/org.talend.sdk.component.studio-integration/src/main/java/org/talend/sdk/component/studio/metadata/migration/tcompv0/JiraTcompv0ToTCKMigrationTask.java (hunk 3 above)
Comment (dmytro-sylaiev) [reply main/plugins/org.talend.sdk.component.studio-integration/src/main/java/org/talend/sdk/component/studio/metadata/migration/tcompv0/JiraTcompv0ToTCKMigrationTask.java]: tacokit component id, now it is in .properties file

File: main/plugins/org.talend.sdk.component.studio-integration/src/main/java/org/talend/sdk/component/studio/util/TcompV0ConfigToFlatMapDeserializer.java (hunk 4 above)
Comment (dmytro-sylaiev) [reply main/plugins/org.talend.sdk.component.studio-integration/src/main/java/org/talend/sdk/component/studio/util/TcompV0ConfigToFlatMapDeserializer.java]: 👍

**Review by mhryb — APPROVED** [review]

File: main/plugins/org.talend.sdk.component.studio-integration/src/main/java/org/talend/sdk/component/studio/metadata/migration/tcompv0/SimpleTcompv0ToTCKMigrationTask.java [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+                                            .filter(key -> key.startsWith(newComponentName) && key.contains("."))
+                                            .forEach((newPropertyKey) -> {
+                                                String tckComponentPropertyName = extractPropertyName(newPropertyKey, newComponentName);
+                                                String componentPropertyType = getComponentPropertyType(migrationProperties, newPropertyKey);
+                                                ComponentUtilities.addNodeProperty(node,
+                                                        tckComponentPropertyName, componentPropertyType);
+                                                String oldValueKey =
+                                                        migrationProperties.getProperty(newPropertyKey);
+                                                String newValue = extractNewPropertyValue(tcompv0PropertiesToMap, oldValueKey);
+                                                ComponentUtilities.setNodeValue(node, tckComponentPropertyName, newValue);
```
Comment (ozhelezniak-talend) [inline main/plugins/org.talend.sdk.component.studio-integration/src/main/java/org/talend/sdk/component/studio/metadata/migration/tcompv0/SimpleTcompv0ToTCKMigrationTask.java:76]: Do we remove old Properties Node in the component configuration?

File: main/plugins/org.talend.sdk.component.studio-integration/src/main/java/org/talend/sdk/component/studio/metadata/migration/tcompv0/JiraTcompv0ToTCKMigrationTask.java [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+            //need a migration for jiraOutput outputAction INSERT->CREATE
+            IComponentFilter filter = new NameComponentFilter("JIRAOutput");
+            try {
+                modified |= ModifyComponentsAction.searchAndModify(item, processType, filter,
+                        Arrays.<IComponentConversion>asList(new IComponentConversion() {
+
+                            public void transform(NodeType node) {
+                                String oldOutputActionValue =
+                                        ComponentUtilities.getNodePropertyValue(node, "configuration.outputAction");
+                                ;
```
Comment (ozhelezniak-talend) [inline main/plugins/org.talend.sdk.component.studio-integration/src/main/java/org/talend/sdk/component/studio/metadata/migration/tcompv0/JiraTcompv0ToTCKMigrationTask.java:43]: ; empty statement

**Review by ozhelezniak-talend — COMMENTED** [review]
left comments

File: main/plugins/org.talend.sdk.component.studio-integration/plugin.xml [hunk 7]
```diff
@@ -314,6 +314,15 @@
            name="UpdateEncryptTckMetadataMigrationTask"
            version="8.0.1">
      </projecttask>
+      <projecttask
+          beforeLogon="false"
+          breaks="8.0.1"
```
[the remark this answers is not in the document]
Comment (dmytro-sylaiev) [reply main/plugins/org.talend.sdk.component.studio-integration/plugin.xml:319]: it is OK, for regular javajet and javajet->tcompv0 breaks and version are often same.

File: main/plugins/org.talend.sdk.component.studio-integration/src/main/java/org/talend/sdk/component/studio/util/TcompV0ConfigToFlatMapDeserializer.java [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    private static void findStoredValueInJson(String initKey, Map<String, String> result, JsonObject value) {
+        JsonValue storedValue = value.get("storedValue");
+        if (value.containsKey("@type") &&
+                "org.talend.daikon.properties.property.EnumProperty"
+                        .equals(String.valueOf(value.getString("@type")))) {
+            JsonObject storedValueJson = storedValue.asJsonObject();
+            String enumStringValue = String.valueOf(storedValueJson.getString("name"));
+            result.put(initKey, enumStringValue);
+        } else if (storedValue != null && JsonValue.ValueType.STRING.equals(storedValue.getValueType())) {
+            result.put(initKey, value.getString("storedValue"));
```
Comment (ozhelezniak-talend) [inline main/plugins/org.talend.sdk.component.studio-integration/src/main/java/org/talend/sdk/component/studio/util/TcompV0ConfigToFlatMapDeserializer.java:38]: What if the type is Boolean?

File: main/plugins/org.talend.sdk.component.studio-integration/src/main/java/org/talend/sdk/component/studio/metadata/migration/tcompv0/SimpleTcompv0ToTCKMigrationTask.java [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+
+                                public void transform(NodeType node) {
+                                    renameTcompV0Component(node, newComponentName);
+                                    setTckVersion(node, newComponentName, migrationProperties);
+
+
+                                    Map<String, String> tcompv0PropertiesToMap = getTcompv0ComponentConfigAsMap(node);
+
+                                    migrationProperties.stringPropertyNames().stream()
+                                            .filter(key -> key.startsWith(newComponentName) && key.contains("."))
```
[the remark this answers is not in the document]
Comment (dmytro-sylaiev) [reply main/plugins/org.talend.sdk.component.studio-integration/src/main/java/org/talend/sdk/component/studio/metadata/migration/tcompv0/SimpleTcompv0ToTCKMigrationTask.java]: Good catch, fixed.
thank you

File: main/plugins/org.talend.sdk.component.studio-integration/src/main/java/org/talend/sdk/component/studio/util/TcompV0ConfigToFlatMapDeserializer.java (hunk 8 above)
Comment (dmytro-sylaiev) [reply main/plugins/org.talend.sdk.component.studio-integration/src/main/java/org/talend/sdk/component/studio/util/TcompV0ConfigToFlatMapDeserializer.java]: same as int or long or double, fixed

File: main/plugins/org.talend.sdk.component.studio-integration/src/main/java/org/talend/sdk/component/studio/metadata/migration/tcompv0/SimpleTcompv0ToTCKMigrationTask.java (hunk 5 above)
Comment (dmytro-sylaiev) [reply main/plugins/org.talend.sdk.component.studio-integration/src/main/java/org/talend/sdk/component/studio/metadata/migration/tcompv0/SimpleTcompv0ToTCKMigrationTask.java:78]: it is removed somehow automatically after migration, but I add the explicit node remove.

File: main/plugins/org.talend.sdk.component.studio-integration/src/main/java/org/talend/sdk/component/studio/metadata/migration/tcompv0/SimpleTcompv0ToTCKMigrationTask.java [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+import org.talend.designer.core.model.utils.emf.talendfile.ElementParameterType;
+import org.talend.designer.core.model.utils.emf.talendfile.NodeType;
+import org.talend.designer.core.model.utils.emf.talendfile.ProcessType;
+import org.talend.sdk.component.studio.util.TcompV0ConfigToFlatMapDeserializer;
+
+/**
+ * The tcompv0->tck migration for case when there's no previous tcompv0->tcompv0 migration needed (post_deserialize).
+ * The .prop
+ */
+public abstract class SimpleTcompv0ToTCKMigrationTask extends AbstractJobMigrationTask {
```
[the remark this answers is not in the document]
Comment (ozhelezniak-talend) [reply main/plugins/org.talend.sdk.component.studio-integration/src/main/java/org/talend/sdk/component/studio/metadata/migration/tcompv0/SimpleTcompv0ToTCKMigrationTask.java:30]: agree to disagree

**Review by ozhelezniak-talend — APPROVED** [review]
Looks ok

**Review by ypiel-talend — APPROVED** [review]
LGTM

**Review by mhryb — APPROVED** [review]

**Pull request merged**